feature/data-migration

SofiaSeniv opened the pull request:
Data migration from json file to db with hosts and their apartments

SofiaSeniv committed:
feat: add data migration

Data migration from json file to db with hosts and their apartments

sawabond commented on BookingService/Migration/Program.cs:
```diff
@@ -0,0 +1,62 @@
+                .ConfigureServices((context, services) =>
+                {
+                    services.AddDbContext<AppDbContext>(options =>
+                        options.UseSqlServer("Data Source=DESKTOP-VM12JTU;Initial Catalog=BookingService;Integrated Security=True;TrustServerCertificate=True;Encrypt=False"));
```
Get from appsettings or console arguments

sawabond commented on BookingService/Migration/Program.cs:
```diff
@@ -0,0 +1,62 @@
+        }
+        catch (Exception ex)
+        {
+            Console.WriteLine($"An error occurred: {ex.Message}");
```
Use ILogger interface. May be Serilog library will be good for it. because we might want logs to be not only in console but in a file or log storage

sawabond commented on BookingService/Migration/Program.cs:
```diff
@@ -0,0 +1,62 @@
+
+        try
+        {
+            var migrationService = host.Services.GetRequiredService<IDataMigrationService>();
```
This way services are not disposed. use something like `using ... CreateScope(); scope.GetRequiredService`

sawabond commented on BookingService/Migration/Program.cs:
```diff
@@ -0,0 +1,62 @@
+        try
+        {
+            var migrationService = host.Services.GetRequiredService<IDataMigrationService>();
+            string filePath = "Files/data.json";
```
From console args or appsettings

sawabond commented on BookingService/Migration/Service/DataMigrationService.cs:
```diff
@@ -0,0 +1,89 @@
+            }
+        }
+
+        private async Task<Result> UserDataMigration(string data)
```
method cannot be named with a noun. Use verbs for method names - `MigrateUserData` for example

sawabond commented on BookingService/Migration/Service/DataMigrationService.cs:
```diff
@@ -0,0 +1,89 @@
+                return Result.Failure("There is no data in file");
+            }
+
+            foreach (var hostData in hostsData)
+            {
+                var host = _mapper.Map<User>(hostData);
+                var creationResult = await _userManager.CreateAsync(host, DefaultPassword);
+                if (!creationResult.Succeeded)
+                {
+                    return Result.Failure("Failure with user migration");
+                }
+                var addToRoleResult = await _userManager.AddToRoleAsync(host, Roles.Host);
+                if (!addToRoleResult.Succeeded)
+                {
+                    return Result.Failure("Failure with user adding to role");
+                }
+
+                foreach (var apartmentData in hostData.Apartments)
+                {
+                    var apartment = _mapper.Map<Apartment>(apartmentData);
+                    apartment.Host = host;
+                    await _apartmentRepository.CreateAsync(apartment);
+                }
+            }
```
Could be optimized. discuss in TG

SofiaSeniv committed:
feat: add logging

SofiaSeniv committed:
refactor: optimize user data migration

SofiaSeniv committed:
refactor: Program.cs file

move strings to appsettings.json file
make services to be disposed

SofiaSeniv committed:
refactor: changes to database schema

Change constraints and OnDelete behaviour
Migration with adding new columns to table

SofiaSeniv committed:
refactor: optimize user data migration removing one extra query to db

Add UserRoles as navigation property
